### ISSUE TITLE
Gapless playback of mp3s

### ISSUE THREAD

**Issue by ThomS8312** [mpv-player/mpv#1101]
Am I overlooking an option in the man page to accomplish this? The decoder I'm using in mpv is the default mpg123.

Gapless playback occurrs in Qmmp if the mpg123 decoder plugin from the Qmmp Plugin Pack is installed and when playing mp3s using mpg123 itself (gapless option is on by default).

I'm at a loss for why it isn't working in mpv.

**Comment by wm4** [comment]
I stared at it again, and it turns out that it's impossible. The Xing VBR seektable simply isn't precise enough (like all mp3 things, it's a terrible and approximate hack that can't facilitate exact seeking).

The only way to implement this would be indexing the file. Either in advance, or during seeking by reading the file completely until the seek target is reached. Both are not so great. I think music players often fix this by scanning the file and then storing an index created from it in their permanent playlist/whatever. But this is not an option for mpv.

Does seeking + correct gapless behavior actually work with mpg123 (the command line tool)? Maybe there's still something that could be done to make it work "well enough", though I doubt it.

**Comment by ThomS8312** [comment]
That's bad news, but maybe something can still be done. Yes, the mpg123 command line tool works as desired. If you use the command <code>mpg123 -C --loop -1 FILE</code>, you can test this out with the audio files I sent a while back. It will loop a file infinitely and with control options. Press <code>h</code> when the file is playing to see all control options.

When traversing my test mp3s using the controls <code>,</code>or<code>;</code> (rewind/fast rewind) and <code>.</code>or<code>:</code> (forward/fast forward), the gapless looping is still flawless.

**Comment by wm4** [comment]
Testing this way with libmpg123 is a bit inconvenient. (I notice this "gap" only when I look at the PCM data in audacity.)

**Comment by wm4** [comment]
PS: do the samples you test have an id3v2 tag at the start of the file?

My own test sample (which I probably got from you, I'm not sure) didn't, and apparently still fails the test, though.

**Comment by ThomS8312** [comment]
<blockquote>My own test sample (which I probably got from you, I'm not sure) didn't, and apparently still fails the test, though.</blockquote>
The sample file I sent has a 15 sec duration. It doesn't have tags, but the result is the same when an id3v2 tag is added. Mpg123 plays the file gaplessly as desired. No pauses or popping sound during the transition is heard at all. It has a smooth transition.

What do you mean fails the test? By my ears, mpg123 command line works just as well as deadbeef/foobar2000 with gapless playback, even when rewinding/forwarding.

**Comment by wm4** [comment]
>What do you mean fails the test? 

I meant your sample file fails the gapless test (as you claim); I was just wondering whether id3v2 tags make it much worse since the ffmpeg seeking code doesn't seem to account for the id3v2 offset.

**Comment by ThomS8312** [comment]
Oh, I misunderstood. You mean mpv.

Just did a test with mpv, and when the file had no tags, it looped without issue. But with an id3v2 tag, I encountered the gapless problem.

**Comment by ThomS8312** [comment]
Just to clarify, the test file loops gaplessly with or without tags if played straight through in mpv. But when seeking, the test file has an audible gap when it has an id3v2 tag. Without the tag, it remains gapless no matter where I seek.

If I were to take a guess, the problem has to do with ffmpeg not taking the offset into account, as you've previously mentioned.

When speaking of id3v2, does that include both id3v2.3 and id3v2.4?

**Comment by wm4** [comment]
Does this ffmpeg patch help? http://sprunge.us/AXgJ

>If I were to take a guess, the problem has to do with ffmpeg not taking the offset into account, as you've previously mentioned.

Yes, this is what the patch fixes.

>When speaking of id3v2, does that include both id3v2.3 and id3v2.4?

Both should be affected, as long as they're on the file start.

**Comment by ThomS8312** [comment]
I'm having a problem patching the file for some reason. I'm getting a "can't find file to patch at input line 5" error.

<blockquote>Both should be affected, as long as they're on the file start.</blockquote>
In what scenerio wouldn't this be the case?

**Comment by wm4** [comment]
I sent the patch to ffmpeg-devel, so you can wait until it's applied.

>In what scenerio wouldn't this be the case? 

In theory id3v2 tags can be at the end of the file too.

**Comment by wm4** [comment]
It was merged into ffmpeg git master just now.

**Comment by ThomS8312** [comment]
The problem is still present after using ffmpeg git. I think it's safe to conclude that the tag is the issue because without the tag, the gapless transition is perfect. Maybe the tag info actually is stored at the end of the file?

The program I use for tagging is puddletag. It could possibly be tagging mp3s in a way that the patch doesn't address.

**Comment by ThomS8312** [comment]
Sent you two files; one with an id3v2.3 tag and the other an id3v2.4 tag. That's to verify if both types of tags are similar or not. If they aren't, they would need separate support.

**Comment by wm4** [comment]
A patch which hopefully fixes this just landed in ffmpeg. Can you try again with ffmpeg git master?

**Comment by ThomS8312** [comment]
I still encountered the problem during my tests, unfortunately. It's as if adding tags causes gapless transition to regress; like tags break something.

In case there's more to this issue than just id3v2 tags, see other possibilities here:
<url>http://wiki.hydrogenaud.io/index.php?title=Gapless_playback#Compression_scheme_artifacts</url>
<url>http://wiki.hydrogenaud.io/index.php?title=Gapless_playback#Poorly_designed_playback_systems</url>

Additional info:
<url>http://wiki.hydrogenaud.io/index.php?title=Gapless_playback#Common_encoder_delay_values</url>
<url>http://wiki.hydrogenaud.io/index.php?title=MP3#VBRI.2C_XING.2C_and_LAME_headers</url>

**Comment by wm4** [comment]
Can you upload tagged and untagged files again and their md5 sums? Are you absolutely sure you tried ffmpeg git master?

**Comment by wm4** [comment]
Another thing to try: ``--demuxer-lavf-o=usetoc=1``.

**Comment by wm4** [comment]
I sent some more patches to ffmpeg, and they were applied just now. The problem was that seeking with CBR was slightly broken, and the size of the tags made it do the wrong thing.

**Comment by ThomS8312** [comment]
Still issues I'm afraid...but I did some more testing after updating and there have been some new discoveries and changes. I will note them in detail.

Both with/without tags, there are no problems when playing straight through or when pausing/unpausing.

When using the mouse cursor and clicking to a playback position, most of the time an audible "click" is present when the file loops. This occurs both with/without tags. Sometimes the "click" isn't present.

Seeking using the mouse wheel and left/right arrows results in an audible "click" <b>and</b> a brief pause just before it while looping. This occurs more so when only seeking forward (scroll mouse wheel up, press right arrow key). If seeking backward (scroll down/left arrow key) after seeking forward, the pause might not be present. Again, this happens both with/without tags. I was unable to test the up/down arrow keys because they exceed the file's duration, but I'd assume the result would be similar.

<blockquote>Can you upload tagged and untagged files again and their md5 sums? Are you absolutely sure you tried ffmpeg git master?</blockquote>
Sent you the files including md5 sums. Yeah, I'm sure. I'm using this source: git://source.ffmpeg.org/ffmpeg.git. Compiled ffmpeg git and mpv git against it to make sure there wouldn't be compatibility issues.

<blockquote>Another thing to try: --demuxer-lavf-o=usetoc=1.</blockquote>
Didn't seem to have had any effect during testing.

**Comment by wm4** [comment]
You said it works with giles without tags, but not with files that have tags. Is this still the case?

**Comment by ThomS8312** [comment]
That's no longer the case after updating to ffmpeg git. That's what I meant when I mentioned "new changes". The "new discoveries" were referring to the mouse wheel/directional arrow key way of seeking. Until the last post, I used the point-and-click way on the OSC to seek through the file (that's how I usually interact with mpv). It's when playing the files through the terminal that I noticed the behavior was different between the mousewheel/directional keys and clicking with the mouse. I double checked to make sure, and have a few more things to note.

I used mpv 0.8.3 and ffmpeg 2.6.1 when I wrote this post:
<blockquote>Just to clarify, the test file loops gaplessly with or without tags if played straight through in mpv. But when seeking, the test file has an audible gap when it has an id3v2 tag. Without the tag, it remains gapless no matter where I seek.</blockquote>
I installed those versions again and got the same result. Because I discovered seeking with the mouse wheel/directional arrow keys can give a different result, I tested that this time too. Often, there's an audible "click" or a slight pause followed by a "click" when seeking that way. The behavior is similar to ffmpeg git. Occurs both with/without tags.

Also, I managed to glitch the OSC when seeking right before the file looped by clicking with the mouse (did it a few times). It freezes the OSC on the screen, but the player still responds. I can no longer seek using a mouse and the OSC doesn't disappear after a couple seconds anymore. I haven't had that happen before with other graphical audio programs I've used that could play mp3s gaplessly.

In my last post, I was using mpv git and ffmpeg git. With the new changes, a "click" is present when looping most of the time, regardless of tags or seeking method. The tags might not be causing the problem then, even though there were problems found and corrected. I now think maybe the issue was just disguised a bit more in ffmpeg stable (2.6.1) than git.

I wonder if the issue has to do with how mpv closes/opens output steams? The glitch I encountered could be a symptom of that. Those hydrogenaudio links I posted before might help to understand the situation better.

**Comment by wm4** [comment]
But you said it works in some cases? When?

>I wonder if the issue has to do with how mpv closes/opens output steams?

We've been over this months ago.

**Comment by ThomS8312** [comment]
Right, I forgot we tried that some time ago.

Testing this issue is difficult because, as I've learned, the results keep changing. Sometimes the looping is gapless, other times there are issues. This is why I've redone my testing and taken a lot more time with it. It's also why I use the term "can" for some of the findings below because it doesn't occur 100% of the time. Here are the results:

FFmpeg 2.6.1 no tag
- Playing without seeking results in no issues.
- Seeking by pointing and clicking anywhere on the progress bar using the <b>mouse cursor</b> results in no issues while looping.
- Seeking forward only (without seeking backward) can result in a "click" sound. This happens using the <b>left arrow key</b> or <b>scrolling up the mouse wheel</b>.
- Seeking in a combination of forward and backward, with the last action being backward, results in no issues while looping when using the <b>left/right arrow keys, mouse wheel,</b> or <b>both</b>.

FFmpeg 2.6.1 tagged
- Playing without seeking results in no issues.
- Seeking by pointing and clicking anywhere on the progress bar using the <b>mouse cursor</b> can result in a "click" sound while looping.
- Seeking forward only (without seeking backward) can result in a "click" sound when using the <b>left arrow key</b> or <b>scrolling up the mouse wheel</b>.
- Seeking in a combination of forward and backward, with the last action being backward, can result in a "click" sound when using the <b>left/right arrow keys, mouse wheel,</b> or <b>both</b>.

FFmpeg git no tag
- Playing without seeking results in no issues.
- Seeking by pointing and clicking anywhere on the progress bar using the <b>mouse cursor</b> results in no issues while looping.
- Seeking forward only (without seeking backward) can result in a "click" sound. This happens using the <b>left arrow key</b> or <b>scrolling up the mouse wheel</b>.
- Seeking in a combination of forward and backward, with the last action being backward, results in no issues while looping when using the <b>left/right arrow keys, mouse wheel,</b> or <b>both</b>.

FFmpeg git tagged
- Playing without seeking results in no issues.
- Seeking by pointing and clicking anywhere on the progress bar using the <b>mouse cursor</b> can result in a "click" sound while looping.
- Seeking forward only (without seeking backward) can result in a "click" sound when using the <b>left arrow key</b> or <b>scrolling up the mouse wheel</b>.
- Seeking in a combination of forward and backward, with the last action being backward, can result in a "click" sound when using the <b>left/right arrow keys, mouse wheel,</b> or <b>both</b>.

Other findings
- Seeking using the mouse near the end of the duration can freeze the OSC (with/without tags).
- The "click" sound can be more like a "crackle" when playing the tagged file.
- A pause can preceed the "click" sound (with/without tags), but seemed more obvious with the tagged file at times.

The results for ffmpeg 2.6.1 and git were identical.

**Comment by ThomS8312** [comment]
<blockquote>Seeking forward only (without seeking backward) can result in a "click" sound. This happens using the left arrow key or scrolling up the mouse wheel.</blockquote>

<blockquote>Seeking forward only (without seeking backward) can result in a "click" sound when using the left arrow key or scrolling up the mouse wheel.</blockquote>

Was supposed to be <b>right arrow key</b>, not left arrow key. Corrected the post.

**Comment by wm4** [comment]
The "click" is on looping long after the seek, not on the seek itself, right?

**Comment by wm4** [comment]
Also, how does it behave with this option set: ``--hr-seek-demuxer-offset=99``.
Seeking will be slower (the higher the seek target the slower), but it should work. Don't try with files longer than 99 seconds.

**Comment by wm4** [comment]
Well, I conclude that it can't be implemented without scanning the full file or so. I'm not exactly sure how other players do it. mpg123 for example (which was confirmed to work) seems to trust the xing tables, which in my opinion can't even work.

**Comment by ThomS8312** [comment]
Tested with <code>--hr-seek-demuxer-offset=99</code>, and made a comparsion to my previous test:

FFmpeg 2.6.1 no tag
- Playing without seeking results in no issues. <b>(Same.)</b>
- Seeking by pointing and clicking anywhere on the progress bar using the mouse cursor results in no issues while looping. <b>(Same.)</b>
- Seeking forward only (without seeking backward) can result in a "click" sound. This happens using the right arrow key or scrolling up the mouse wheel. <b>(Same.)</b>
- Seeking in a combination of forward and backward, with the last action being backward, results in no issues while looping when using the left/right arrow keys, mouse wheel, or both. <b>(Same.)</b>

FFmpeg 2.6.1 tagged
- Playing without seeking results in no issues. <b>(Same.)</b>
- Seeking by pointing and clicking anywhere on the progress bar using the mouse cursor can result in a "click" sound while looping. <b>(The "clicking" has reduced considerably, but still happens.)</b>
- Seeking forward only (without seeking backward) can result in a "click" sound when using the right arrow key or scrolling up the mouse wheel. <b>(Same.)</b>
- Seeking in a combination of forward and backward, with the last action being backward, can result in a "click" sound when using the left/right arrow keys, mouse wheel, or both.<b>(Same.)</b>

FFmpeg git no tag
- Playing without seeking results in no issues. <b>(Same.)</b>
- Seeking by pointing and clicking anywhere on the progress bar using the mouse cursor results in no issues while looping. <b>(Same.)</b>
- Seeking forward only (without seeking backward) can result in a "click" sound. This happens using the right arrow key or scrolling up the mouse wheel. <b>(Same.)</b>
- Seeking in a combination of forward and backward, with the last action being backward, results in no issues while looping when using the left/right arrow keys, mouse wheel, or both. <b>(Same.)</b>

FFmpeg git tagged
- Playing without seeking results in no issues. <b>(Same.)</b>
- Seeking by pointing and clicking anywhere on the progress bar using the mouse cursor can result in a "click" sound while looping. <b>(The "clicking" has reduced considerably, but still happens.)</b>
- Seeking forward only (without seeking backward) can result in a "click" sound when using the right arrow key or scrolling up the mouse wheel. <b>(Same.)</b>
- Seeking in a combination of forward and backward, with the last action being backward, can result in a "click" sound when using the left/right arrow keys, mouse wheel, or both. <b>(Same.)</b>

The results for ffmpeg 2.6.1 and git were identical again.

<blockquote>The "click" is on looping long after the seek, not on the seek itself, right?</blockquote>
I haven't noticed "clicking" with the seek itself, just during looping.

Is it confirmed that mpg123 uses the "xing tables" method?

**Comment by wm4** [comment]
Testing ffmpeg 2.6.1 is not really useful. What ffmpeg git version did you test?

I'm pretty convninced that there should be no difference between tagged and no-tagged files, but In could be wrong.

Thanks for all the testing. One thing: summarizing the results in a table or so would be better. I always have to glance over the whole post to find the differences.

**Comment by wm4** [comment]
>Is it confirmed that mpg123 uses the "xing tables" method?

It does have code for this, but after closer inspection and running mpg123 in debug mode, it seems that it actually parses and skips mp3 frames, which effectively indexes the file until the target position is reached. (The mpg123 code is a bit hard to read...) This probably leads to rather bad performance when seeking very large files, but after all it's the only reasonable way to handle this. I'll look into implementing this in ffmpeg, probably.

**Comment by wm4** [comment]
> I'll look into implementing this in ffmpeg, probably.

Well, I'll definitely do that. No need to do any further tests until that's done.

**Comment by ThomS8312** [comment]
<blockquote>I'm pretty convninced that there should be no difference between tagged and no-tagged files, but In could be wrong.</blockquote>
Hm, maybe in theroy. When I tested this out though, I played it over and over ad nauseum to get as many results as I could in case I missed something due to its dynamic nature. If you still have the files, you can try it out and compare the results.

<blockquote>Thanks for all the testing. One thing: summarizing the results in a table or so would be better. I always have to glance over the whole post to find the differences.</blockquote>

No problem. I spent a good part of the last couple of days experimenting with seeking. Time consuming, but worth it in helping to solve the problem. I made a table of the last tests and will post it soon. I also removed the results of ffmpeg 2.6.1 and added the git version.

Btw, the git version I've been using was after you made some additional updates because, "CBR was slightly broken, and the size of the tags made it do the wrong thing."

<blockquote>It does have code for this, but after closer inspection and running mpg123 in debug mode, it seems that it actually parses and skips mp3 frames, which effectively indexes the file until the target position is reached. (The mpg123 code is a bit hard to read...) This probably leads to rather bad performance when seeking very large files, but after all it's the only reasonable way to handle this.</blockquote>

This sounded really familiar to me, and I remembered why. Earlier I mentioned that foobar2000 played mp3s flawlessly (which is what deadbeef is modeled after and plays mp3s flawlessly as well).

Take a look at this: <url>http://www.foobar2000.org/FAQ</url>. Go to the 4th question in the troubleshooting section, "Why is seeking so slow while playing MP3 files?" The answer sounds similar to what you found after examining mpg123 in debug mode. I remember reading that ages ago...

**Comment by ThomS8312** [comment]
<pre><code>|--------------------------|---------------|---------------|
|FFmpeg                    |               | <b>"Clicking"</b>    |
|Git Ver. r71486.4efc0e6   |               | <b>with option:</b>  |
|Without ID3v2.3 Tag       |               | <b>--hr-seek-</b>    |
|                          |               |   <b>demuxer-</b>    |
| <b>Seeking Method</b>           | <b>"Clicking"?</b>   |   <b>offset=99?</b>  |
|--------------------------|---------------|---------------|
| No Seeking               |      No       |      No       |
|--------------------------|---------------|---------------|
| Mouse Cursor             |      No       |      No       |
|--------------------------|---------------|---------------|
| Arrow Keys: Right        |               |               |
| Mouse Wheel: Up          |      Yes      |      Yes      |
| (Forward Only)           |               |               |
|--------------------------|---------------|---------------|
| Arrow Keys: Right + Left |               |               |
| Mouse Wheel: Up + Down   |      No       |      No       |
| <i>(Forward & Backward)**</i>   |               |               |
|--------------------------|---------------|---------------|


|--------------------------|---------------|---------------|
|FFmpeg                    |               | <b>"Clicking"</b>    |
|Git Ver. r71486.4efc0e6   |               | <b>with option:</b>  |
|With ID3v2.3 Tag          |               | <b>--hr-seek-</b>    |
|                          |               |   <b>demuxer-</b>    |
| <b>Seeking Method</b>           | <b>"Clicking"?</b>   |   <b>offset=99?</b>  |
|--------------------------|---------------|---------------|
| No Seeking               |      No       |      No       |
|--------------------------|---------------|---------------|
| Mouse Cursor             |      Yes      |      <i>Yes*</i>     |
|--------------------------|---------------|---------------|
| Arrow Keys: Right        |               |               |
| Mouse Wheel: Up          |      Yes      |      Yes      |
| (Forward Only)           |               |               |
|--------------------------|---------------|---------------|
| Arrow Keys: Right + Left |               |               |
| Mouse Wheel: Up + Down   |      Yes      |      Yes      |
| <i>(Forward & Backward)**</i>   |               |               |
|--------------------------|---------------|---------------|

*  <i>The "clicking" has reduced considerably, but still happens.</i>
** <i>Includes arrow keys and mouse wheel used in combination.</i>

NOTE:
1. Seeking using the mouse near the end of the duration can freeze the OSC (with/without tags).
2. The "click" sound can be more like a "crackle" when playing the tagged file.
3. A pause can preceed the "click" sound (with/without tags), but seemed more obvious with the tagged file at times.
</code></pre>

**Comment by wm4** [comment]
>1. Seeking using the mouse near the end of the duration can freeze the OSC (with/without tags).

I can't reproduce this, btw.

**Comment by wm4** [comment]
Another change landed in ffmpeg git. Seeking in mp3 is now slower, but should be always exact. Don't use use of the ``usetoc`` options when testing.

**Comment by ThomS8312** [comment]
<blockquote>I can't reproduce this, btw.</blockquote>

I first encountered the bug by just single clicking near the end of the file duration when doing tests on mouse seeking. It's probably easier to make this happen if you hold down the left mouse button and move back and forth near the very end of the file. I just tried it this way and was able to do it 5 times.

I've had the freeze happen in two different ways.
1. The OSC freezes in its normal playing size with the diamond marker at the the last position it had.
2. The player is smaller, has no diamond marker in the seekbar, no time passed and time remaining under the seekbar, and the audio tracks icon is greyed out. You can see the smaller player flash for a second sometimes if you have the OSC visible on the screen while looping (by keeping mouse cursor on the OSC).

**Comment by wm4** [comment]
Probably fixed now.

**Comment by ThomS8312** [comment]
Good news! The last ffmpeg update appears to have corrected the gapless issue. There was no "click" or pause found during looping anymore.

However, after updating mpv git, the player can also crash now when I do seeking near the end of the file's duration and the freeze is still present. I even got a third type of freeze:
3. Freezes at normal size. But the diamond marker, time passed, and time remaining are absent on the OSC.

I did some comparisons between formats and discovered that this freeze issue is not exclusive to mp3s. It happened with m4a and wav files too, but flac files don't seem to be affected. The seeking behaves differently between the mouse wheel/arrow keys and the mouse cursor:

<pre>
<i>Mouse cursor</i> (MP3, M4A, WAV)
Sometimes <b>freezes or crashes</b> when seeking near the end of duration.

<i>Mouse cursor</i> (FLAC)
<b>Did not</b> result in a <b>freeze or crash</b>. If holding down the left mouse button near the very end
of the file, the position will often jump back a few seconds and then reach the file's end.
Otherwise, it looped to the beginning.</pre>

Note: Continuous seeking is done by holding down the right arrow key so that the file plays over and over again.

<pre>
<i>Mouse wheel/arrow keys</i> (MP3, M4A, WAV)
The file <b>continuosly seeks</b> over and over when looping without interruption until there's a
<b>possible crash</b>, but <b>did not freeze</b>.

<i>Mouse wheel/arrow keys</i> (FLAC)
The file <b>did not continously seek</b> over and over when looping without interruption and
<b>did not freeze or crash</b>. I am constantly interrupted when I reach the duration's end.
Sometimes I can get it to continously seek at loop (with a shorter file), but it's only for
a few times. When looping occurs those few times, I noticed (as shown by the on-screen
progress bar) that the file loops at a point that's a few seconds before the file ends.
When continuous seeking doesn't occur (when I'm prevented even though I'm holding down the 
right arrow key) and the end of the file is reached, the position jumps back a few seconds
then ends.</pre>

**Comment by wm4** [comment]
>Good news! The last ffmpeg update appears to have corrected the gapless issue. There was no "click" or pause found during looping anymore.

Good, then I'm closing this.

>Sometimes freezes or crashes when seeking near the end of duration.

I doubt the player actually crashes or freezes.

What you refer to as crash is probably the player exiting. If you seek immediately to the end, the file immediately ends playback after loading. Then the player thinks the file is unplayable and skips it. No more files to play make the player exit. You can stop this behavior with ``--loop=force``.

It shouldn't freeze either. What you see is probably the OSC script crashing. If so, this is a bug that should be fixed.

Can you post terminal output with ``-v`` in both cases?

**Comment by ThomS8312** [comment]
I haven't been able to make the player "crash" after using <code>--loop=force</code>. Sent you logs.

**Comment by wm4** [comment]
I don't know why the OSC crashes and couldn't reproduce it - but I added a sanity check, which might take care of it.

For the record, this is the log:
```
[cplayer] Run command: seek, flags=0, args=[99.112426, relative, keyframes]
[osc] 
[osc] stack traceback:
[osc] 	@osc.lua:454: in function 'render_elements'
[osc] 	@osc.lua:1777: in function 'render'
[osc] 	@osc.lua:1905: in function 'tick'
[osc] 	@osc.lua:1861: in function 'process_event'
[osc] 	@osc.lua:1965: in function 'cb'
[osc] 	mp.defaults:83: in function 'fn'
[osc] 	mp.defaults:42: in function 'handler'
[osc] 	mp.defaults:316: in function 'handler'
[osc] 	mp.defaults:418: in function 'call_event_handlers'
[osc] 	mp.defaults:454: in function 'dispatch_events'
[osc] 	mp.defaults:411: in function <mp.defaults:410>
[osc] 	[C]: in ?
[osc] 	[C]: in ?
[osc] Lua error: @osc.lua:454: attempt to index field 'eventresponder' (a nil value)
[osc] Exiting...
[lavf] Found 'mp3' at score=1 size=2048.
```
So it happens while the file is being loaded.

**Comment by ThomS8312** [comment]
<blockquote>I added a sanity check, which might take care of it.</blockquote>
I'll test it out shortly.

Well, I encountered another scenerio where the gapless issues are back. That behavior returns with embedded album art (which my test file didn't have originally, only the tag). The formats I used for album art were jpg, png, and gif. All were affected.

Also, if multiple images are embedded with a description, the <code>show_text</code> command with the <code>track-list</code> property displays various spacing between the last character in the descriptions and the closing <code>'</code> that encloses the description. For example:

<pre>
Video: > (1) 'Description ' <
Video: (2) 'Description'
Video: (3) 'Description      '</pre>

The second entry is what it should look like. There shouldn't be any space there (unless trailing whitespace were a part of the description itself, but that isn't the case here).

**Comment by wm4** [comment]
Can you send sample files?

**Comment by ThomS8312** [comment]
Sent.

**Comment by wm4** [comment]
Thanks, will look tomorrow.

**Comment by ThomS8312** [comment]
This is interesting:
<pre>
The <b>png</b> embedded file with a single embedded image <b>did not have gapless issues</b> and <b>crashed</b> while seeking.
The <b>png</b> embedded file with multiple embedded images <b>had gapless issues</b> and <b>crashed</b> while seeking.

The <b>jpg</b> embedded file with a single embedded image <b>had gapless issues</b> and <b>did not crash</b> while seeking.
The <b>jpg</b> embedded file with multiple embedded images <b>had gapless issues</b> and <b>did not crash</b> while seeking.

The <b>gif</b> embedded file with a single embedded image <b>had gapless issues</b> and <b>did not crash</b> while seeking.
The <b>gif</b> embedded file with multiple embedded images <b>had gapless issues</b> and <b>did not crash</b> while seeking.</pre>

I checked the original test file too:
<pre>
The original test file with no embedded image(s) <b>did not have gapless issues</b> and <b>crashed</b> while seeking.</pre>

An effective way to cause a crash is by holding down the up arrow key with the option <code>loop=inf</code> until it crashes. The original test file and png embedded files (both single and multiple images) are pretty easy to crash, but the png embedded files can be a bit resilient at times.

**Comment by ThomS8312** [comment]
I found another issue regarding what the <code>track-list</code> property displays using the <code>show_text</code> command. When mp3 files have a single image embedded that has no description, the <code>''</code> that encloses the description are visible, but should not be. When mp3 files have multiple images embedded that have no description, the same strange spacing issue I mentioned earlier is shown. 

Example 1 (single embedded image):
<pre>
Video: > (1) '' <</pre>

Example 2 (multiple embedded images):
<pre>
Video: > (1) '  ' <
Video: (2) '' 
Video: (3) '     '
</pre>

**Comment by wm4** [comment]
The gaps happened because of another ffmpeg issue. If the size of the tags was large enough, it made ffmpeg think this was a concatenated file, and ignored the file duration header in the mp3 file as consequence of it. This in turn disabled the gapless-skipping at the end of the file. I sent a patch to ffmpeg to handle this better, and it was merged. All files should play fine now.

What crashes are you talking about? I haven't seen a single crash yet. The OSC broke in some situations; if you found another case where this happens, please post the Lua backtrace.

Some tag fields have spaces. But they are stored like this in the file, and of course mpv preserves them. You can use e.g. the id3v2 utility to confirm that the relevant fields actually contain spaces:
```
APIC (Attached picture): (Enter description      )[, 3]: image/png, 9951 bytes
APIC (Attached picture): (Enter description          )[, 3]: image/png, 9951 bytes
APIC (Attached picture): (Enter description               )[, 3]: image/png, 9951 bytes
APIC (Attached picture): (Enter description)[, 3]: image/png, 9951 bytes
APIC (Attached picture): (Enter description )[, 3]: image/png, 9951 bytes
APIC (Attached picture): (Enter description   )[, 3]: image/png, 9951 bytes
```
If id3v2 tags like this are common, then maybe they should be stripped, but I'm not convinced of this yet.

**Comment by ThomS8312** [comment]
Looks like the new gapless issue is fixed now.

<blockquote>What crashes are you talking about? I haven't seen a single crash yet. The OSC broke in some situations; if you found another case where this happens, please post the Lua backtrace.</blockquote>

My mistake, I meant "closes" not "crashes". I noted the behavior in case it was a clue to the new gapless problem because I noticed those differences. I edited that post. The OSC issue still occurs though.

<blockquote>Some tag fields have spaces. But they are stored like this in the file, and of course mpv preserves them. </blockquote>
Hm. Id3v2 does in fact show those spaces. Is that how the id3v2 format treats descriptions?

The strange thing is that the <code>''</code> are visible (they shouldn't be without a description) in every single mp3 with an embedded image containing no description that I've tried. It makes me think a non-printable character might be triggering its display and that's why the <code>''</code> shows up when "nothing" is there. If it were a whitespace, there would be a separation between them like so: <code>' '</code>. Maybe those "spaces" are something else?

**Comment by wm4** [comment]
>Hm. Id3v2 does in fact show those spaces. Is that how the id3v2 format treats descriptions?

No. The id3v2 format simply allows the user to enter spaces, even if there are no other characters after the spaces. The strings are variable length, so the spaces are not required padding or anything. So mpv (and probably some other tools) show them like this.

>The strange thing is that the '' are visible (they shouldn't be without a description) in every single mp3 with an embedded image containing no description that I've tried.

None of your sample files were like this.

**Comment by ThomS8312** [comment]
<blockquote>No. The id3v2 format simply allows the user to enter spaces, even if there are no other characters after the spaces. The strings are variable length, so the spaces are not required padding or anything. So mpv (and probably some other tools) show them like this.</blockquote>

I'm not putting the trailing spaces in though, so I'm not sure why they're present.

<blockquote>None of your sample files were like this.</blockquote>

The ones I sent yesterday wouldn't because all the covers had descriptions, and the earlier files I sent wouldn't because they didn't have covers. Sent you files with covers but no description.

**Comment by wm4** [comment]
>I'm not putting the trailing spaces in though, so I'm not sure why they're present.

What tool did you use? Also I guess I'll took a closer look again.

>Sent you files with covers but no description.

OK, looked at them, and I realize that this is because the attachments always have a description, so if none was entered, it's an empty string. But it's not pretty when displyed. I'll think about what's the best way to fix it.

**Comment by ThomS8312** [comment]
I used puddletag to tag the files. Do you think puddletag is injecting those strange spaces?

**Comment by wm4** [comment]
Possibly - I'll try myself tomorrow.

**Comment by wm4** [comment]
Looked at puddletag - it doesn't seem to add extra spaces by itself.

**Comment by ThomS8312** [comment]
That's very strange. If it's not puddletag, why is that happening and how to fix it?

It might be a weird issue with the mp3 format itself because it affects <b>all</b> mp3s I've checked. If an image is embedded in an mp3 file, the spacing problem occurs.

**Comment by wm4** [comment]
Is the difference whether they have a ID3v1 tag or a ID3v2 tag? Since v1 can't have embedded images.

**Comment by ThomS8312** [comment]
It appears that the spacing issue is independent of ID3 tags because it's still present on mp3s without ID3 tags.

**Comment by wm4** [comment]
Without id3 tags? Then what tags do they have?

**Comment by ThomS8312** [comment]
Well, just learned a couple of things. Embedded images are a part of ID3v2 and the files I thought had "no tags" had an ID3v1 tag. I had puddletag set to create one if one wasn't present without realizing it.

I removed both ID3v1 and ID3v2 tags from one test file, then added the embedded images again making sure puddletag wouldn't create an ID3v1 tag. The spacing issue still occurs, but the spacing is different. Both files are identical except for ID3v1 tags:

With ID3v1 tags (no description):
<pre>
APIC (Attached picture): (      )[, 3]: image/jpeg, 62963 bytes
APIC (Attached picture): (   )[, 3]: image/jpeg, 62963 bytes
APIC (Attached picture): (          )[, 3]: image/jpeg, 62963 bytes
APIC (Attached picture): ()[, 3]: image/jpeg, 62963 bytes
APIC (Attached picture): ( )[, 3]: image/jpeg, 62963 bytes
</pre>

Without ID3v1 tags (no description):
<pre>
APIC (Attached picture): (      )[, 3]: image/jpeg, 62963 bytes
APIC (Attached picture): ()[, 3]: image/jpeg, 62963 bytes
APIC (Attached picture): ( )[, 3]: image/jpeg, 62963 bytes
APIC (Attached picture): (          )[, 3]: image/jpeg, 62963 bytes
APIC (Attached picture): (               )[, 3]: image/jpeg, 62963 bytes
APIC (Attached picture): (   )[, 3]: image/jpeg, 62963 bytes
</pre>

I opened the files again in puddletag (after all operations were done), and the various lengthed spaces that correspond to the above output are whitespaces in the description text box within puddletag.

**Comment by wm4** [comment]
It's still not clear to me where the spaces come from...

**Comment by ThomS8312** [comment]
I did a comparison using <code>eyeD3</code> (another ID3 tag editing program) and the spacing issue still occured. Here's a snippet of the explanation to the <code>--add-image</code> option in that program:
<pre>
Add or replace an image. There may be more than one
image in a tag, as long as the DESCRIPTION values are
unique. The default DESCRIPTION is ''.</pre>

The reason we are seeing the spaces is because the description values <b>have to be unique</b>. Some of the sample files I sent with more than one embedded image had the same "Enter description" description or had no description, which is the same situation. Here's the output for multiple embedded images with unique descriptions (by removing letters at the end):
<pre>
APIC (Attached picture): (Enter descripti)[, 3]: image/jpeg, 62963 bytes
APIC (Attached picture): (Enter descriptio)[, 3]: image/jpeg, 62963 bytes
APIC (Attached picture): (Enter descrip)[, 3]: image/jpeg, 62963 bytes
APIC (Attached picture): (Enter description)[, 3]: image/jpeg, 62963 bytes
APIC (Attached picture): (Enter descri)[, 3]: image/jpeg, 62963 bytes
APIC (Attached picture): (Enter descript)[, 3]: image/jpeg, 62963 bytes</pre>

The spaces are gone. It seems the whitespace is automatically added for the sake of making the descriptions unique.

The order of the embedded images are not kept as they are saved either. That output should have been like this if the original order was respected:
<pre>
APIC (Attached picture): (Enter description)[, 3]: image/jpeg, 62963 bytes
APIC (Attached picture): (Enter descriptio)[, 3]: image/jpeg, 62963 bytes
APIC (Attached picture): (Enter descripti)[, 3]: image/jpeg, 62963 bytes
APIC (Attached picture): (Enter descript)[, 3]: image/jpeg, 62963 bytes
APIC (Attached picture): (Enter descrip)[, 3]: image/jpeg, 62963 bytes
APIC (Attached picture): (Enter descri)[, 3]: image/jpeg, 62963 bytes</pre>

Also, the <code>''</code> mentioned in the explanation must be literal. That's what's shown when there are no descriptions specified on files with only one embedded image, making <code>''</code> the <b>default</b> description just as described above. Those "issues" are the normal behavior of ID3v2, but not good behavior IMO. Compared to other multimedia programs, Mpv is honest with ID3v2 descriptions. But I think a little <i>too</i> honest in this case.

I have two questions:
1. Is it possible to distinguish white spaces added by the user at the end of a description and ones that are automatically added to nonunique descriptions?
2. Is there some kind of identifier in the metadata that indicates which image was intended to be first, second, third...etc?

**Comment by wm4** [comment]
Uh, interesting find. Do you know if id3v2 acrually mandates unique descriptions? How do programs behave if they aren't?

The tags obviously have an order, so in theory a tag editor or reader could return the tags in the order they appeared in the file. I'm pretty sure ffmpeg/mpv will do this.

We could perhaps discard the description if it consists only of spaces (or is empty), or strip trailing spaces. For the spaces thing, I'd need to know if it happens only with a specific tag editor, or with more programs.

**Comment by ThomS8312** [comment]
According to <url>http://id3.org/id3v2.3.0#Attached_picture</url>:
<blockquote>
There may be several pictures attached to one file, each in their individual "APIC" frame,
but only one with the same content descriptor.</blockquote>
If "content descriptor" means the description, then it is mandated. I recommend reading the whole section because it provides other information that may be useful.

<code>eyeD3</code> will overwrite an image with a matching description. If it does not match, it is added in addition. <code>eyeD3</code> saves them in the right order, one after the other as long as they are unique.

Puddletag automatically adds spaces to deal with nonunique descriptions, and for some reason, saves multiple images to an mp3 in a different order than was arranged.

Now that I know the reasons behind the behavior, I think the best way to resolve this is to simply document it in the man page. If information about the behavior is available, it can be understood. Adding that info to the<code>Property list</code> --> <code>track-list</code> section is probably a good spot for it since expanding that property with the <code>show_text</code> command is how a user would find out about it (like I did).

Points to cover:
- Pertains to ID3v2 tags.
- Descriptions must be unique.
- A single embedded image with no description creates a <code>''</code> description.
- Adding trailing whitespaces to a description is a solution to the limitation of having unique descriptions. That includes cases when multiple images have no description (<code>''</code>). Tagging software with ID3v2 support may add whitespaces automatically.

**Comment by ThomS8312** [comment]
The concern I have with stripping whitespaces and blank descriptions is that someone may want to have them. The behavior for ID3v2 would be disproportionate to other tag displays. For example, I would be able to have a trailing whitespace displayed in a Vorbis comment, but wouldn't be able to with an ID3v2 tag.

We could go halfway and create an option to discard whitespace, but I think it decides too much for the user if it's implemented without a means to disable it. If an option for hiding empty description strings (<code>''</code>) and trailing whitespaces can be implemented but also disabled, it would be an even better solution than just documentation.

Could this be created as an option?

**Comment by wm4** [comment]
When would you ever use such an option?

**Comment by ThomS8312** [comment]
I picture the option as one that would be added to mpv.conf and takes effect when enabled. It would make the display visually appealing for ID3v2 tags, but just masking over what's really there to make things pretty.

If someone prefers a clean aesthetic over accurate information, they could have the option enabled. If someone wants it the other way around, they could disable it (or not have it added in the first place). Or switch back and forth.

An option offers flexibility without the commitment, because the method used to make the display of ID3v2 tags visually appealing comes at a cost to accurate feedback. A user can see the ugly metadata as it actually is but have the choice of covering it up too.

**Comment by wm4** [comment]
Well, I'd prefer to do "the right thing", and preferably in FFmpeg. And I think it should just trim trailing spaces, because they are being added as a hack for ID3 deficiencies, and have no other use. Would this be ok? (Of course there's still a chance ffmpeg rejects such a patch.)

**Comment by ThomS8312** [comment]
Oh, so you want to change FFmpeg instead of mpv. That changes what I thought originally. Sure, let's see if it'll be accepted.

Yeah, trailing whitespace is indeed being used specifically as a workaround. Besides, having trailing whitespace in general isn't a common practice anyway, so I don't see it really taking away from much (if anything).

My interest was to have ID3 treated equally to other tags, but it really can't due to the limitations. It's a unique situation that needs to be treated as such. ID3 just doesn't have the same ability as, for example, Vorbis comments in this situation.

Do you think it would be practical to add a message about stripping the trailing whitespaces and empty strings (<code>''</code>) of ID3v2 tags during a diagnostic scenerio? It would just be a way to inform users of what's happening in a setting where the user requests output, like verbose.

**Comment by wm4** [comment]
>My interest was to have ID3 treated equally to other tags, but it really can't due to the limitations. It's a unique situation that needs to be treated as such. ID3 just doesn't have the same ability as, for example, Vorbis comments in this situation.

Yes... and such things do happen often. IMO ffmpeg shouldn't try to accommodate every semi-broken format perfectly, and it's ok if sometimes information is not perfectly exported.

**Comment by ThomS8312** [comment]
What do you think of having a message displayed about the trailing whitespace/empty strings?

**Comment by wm4** [comment]
>What do you think of having a message displayed about the trailing whitespace/empty strings?

Hm... not much.

I've sent 2 patches to ffmpeg: 1. drop empty picture titles in id3v2 tags, and 2. strip trailing spaces from id3v2 tags. Patch 1 was applied, but patch 2 was silently ignored so far.

**Comment by ThomS8312** [comment]
I suppose it's not absolutely needed, but as a user I do appreciate readily available information about a program's behavior (especially when it change's something without user input).

I discovered yesterday that puddletag actually does include information about the whitespaces being appended to ID3 descriptions. It's conviently displayed as a hint when you hover over the description text box.

**Comment by wm4** [comment]
If nothing will change on the ffmpeg side, I might add a custom hack to trim APIC titles.

**Comment by ThomS8312** [comment]
The ffmpeg guys are really dragging their feet on this one...but I think that's a good thing. Their silence kind of validates my earlier concern.

I say it's better for mpv to have the authority to strip the whitespaces in descriptions instead of ffmpeg. Documenting that behavior would also be less of a hassle, more worthwhile, and easily seen by the user with a mention in the man page.

**Comment by wm4** [comment]
>The ffmpeg guys are really dragging their feet on this one...but I think that's a good thing. Their silence kind of validates my earlier concern.

Uh, what? Nobody has been even thinking about this for 1 month.

**Comment by ThomS8312** [comment]
Yeah, that's what I mean. We were waiting on ffmpeg to accept the patch to strip trailing whitespace, and it's a month later at this point. They haven't accepted it...or have they?

**Comment by wm4** [comment]
One patch was merged, one wasn't (without comment, but I've sent the patch together with a comment that the patch is questionable and I wouldn't mind if it's not applied).

I can send it again if you want.

**Comment by ThomS8312** [comment]
<blockquote>One patch was merged, one wasn't</blockquote>

The patch that was merged was for the removal of displaying <code>''</code> when there's no description present. That change is not problematic because it doesn't interfere with user/program input. However, the one that wasn't merged does.

<blockquote>I can send it again if you want.</blockquote>

Well, I think a custom hack for mpv would actually be better. I feel it's more appropriate for mpv to remove trailing whitespace instead of ffmpeg because it's less aggressive that way.

The means to document that behavior is also more practical if mpv is responsible for it. Adding an explanation of why mpv removes trailing whitespace of ID3 descriptions to the man page is an easy enough task.

**Comment by wm4** [comment]
>Well, I think a custom hack for mpv would actually be better. I feel it's more appropriate for mpv to remove trailing whitespace instead of ffmpeg because it's less aggressive that way.

Normally, I don't want to do this, because
1. inside of FFmpeg is the better place, because it's the correct place for format specifics,
2. such hacks just keep piling up

**Comment by wm4** [comment]
I asked again, and they applied the patch.

**Comment by ThomS8312** [comment]
Since that's the case, I agree it would be proper to include it in ffmpeg and a patch should be sent again. Maybe what I'm trying to accomplish is unusual.

I'm trying to prevent confusion for when a program behaves as an authority in a way that a user may not understand. That's why I mentioned documentation. I can see a situation where a user is able to view trailing whitespace with another program (ie. tag editor), but when using ffmpeg/mpv see that it's not present and mistake the difference as a bug.

Is it practical to have the user informed? If a user wanted to find out why the trailing whitespace is removed, how would they find out?

**Comment by ThomS8312** [comment]
Well, that was quick. I got your message right as I sent mine. I'm glad they accepted.
